Mobile Buy SDK/Mobile Buy SDK/Operations/BUYCheckoutOperation.m

```diff
@@ -0,0 +1,168 @@
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+#import "BUYCheckoutOperation.h"
+#import "BUYClient+Checkout.h"
+#import "BUYPaymentToken.h"
+#import "BUYCheckout.h"
+#import "BUYRequestOperation.h"
+
+@interface BUYClient (PrivateCheckout)
+
+- (BUYRequestOperation *)beginCheckout:(BUYCheckout *)checkout paymentToken:(id<BUYPaymentToken>)paymentToken completion:(BUYDataCheckoutBlock)block;
+- (BUYRequestOperation *)getCompletionStatusOfCheckoutToken:(NSString *)token start:(BOOL)start completion:(BUYDataStatusBlock)block;
+- (BUYRequestOperation *)getCheckout:(BUYCheckout *)checkout start:(BOOL)start completion:(BUYDataCheckoutBlock)block;
+
+- (void)startOperation:(BUYOperation *)operation;
+
+@end
+
+@interface BUYCheckoutOperation ()
+
+@property (strong, nonatomic, readonly) BUYCheckout *checkout;
+@property (strong, nonatomic, readonly) id<BUYPaymentToken> token;
+@property (strong, nonatomic, readonly) BUYCheckoutOperationCompletion completion;
+
+@property (strong, nonatomic) BUYRequestOperation *currentOperation;
+
+@end
+
+@implementation BUYCheckoutOperation
+
+#pragma mark - Init -
+
++ (instancetype)operationWithClient:(BUYClient *)client checkout:(BUYCheckout *)checkout token:(id<BUYPaymentToken>)token completion:(BUYCheckoutOperationCompletion)completion
+{
+	return [[[self class] alloc] initWithClient:client checkout:checkout token:token completion:completion];
+}
+
+- (instancetype)initWithClient:(BUYClient *)client checkout:(BUYCheckout *)checkout token:(id<BUYPaymentToken>)token completion:(BUYCheckoutOperationCompletion)completion
+{
+	self = [super init];
+	if (self) {
+		_client     = client;
+		_checkout   = checkout;
+		_token      = token;
+		_completion = completion;
+	}
+	return self;
+}
+
+#pragma mark - Finishing -
+
+- (void)finishWithCheckout:(BUYCheckout *)checkout
+{
+	[self finishExecution];
+	self.completion(checkout, nil);
+}
+
+- (void)finishWithError:(NSError *)error
+{
+	[self finishExecution];
+	self.completion(nil, error);
+}
+
+#pragma mark - Execution -
+
+- (void)startExecution
+{
+	if (self.isCancelled) {
+		return;
+	}
+	
+	[super startExecution];
+	
+	__weak typeof(self) weakSelf = self;
+	
+	BUYRequestOperation *beginOperation = [weakSelf.client beginCheckout:weakSelf.checkout paymentToken:weakSelf.token completion:^(BUYCheckout *checkout, NSError *error) {
+		if (weakSelf.isCancelled) {
+			return;
+		}
+		
+		if (!checkout) {
+			[weakSelf finishWithError:error];
+			return;
+		}
+		
+		BUYRequestOperation *pollOperation = [weakSelf.client getCompletionStatusOfCheckoutToken:weakSelf.checkout.token start:NO completion:^(BUYStatus status, NSError *error) {
```

Comment:
Something that concerns me here: the completion blocks of one operation are referring to the original operation. It looks like most of this code would not execute, because the original operation is going to get removed from the operation queue. It's implicit that whoever creates the first operation has to retain a strong reference to it, otherwise it will be deallocated and the subsequent operations will not happen. Am I wrong here? I think we might have to avoid using references to the operations in the completion blocks. For example, by declaring strong local variables, or perhaps by using properties of the subsequent operations instead of the originating one.